sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala

```diff
@@ -187,6 +187,16 @@ case object PREFER_SHUFFLE_HASH extends JoinStrategyHint {
   override def hintAliases: Set[String] = Set.empty
 }
 
+/**
+ * An internal hint to prohibit broadcasting and replicating one side of a join. This hint is used
+ * by some rules where broadcasting or replicating a particular side of the join is not permitted,
+ * such as the MERGE cardinality check.
+ */
+case object NO_BROADCAST_AND_REPLICATION extends JoinStrategyHint {
```

Comment:
Shall we spin off this `NO_BROADCAST_AND_REPLICATION` as an independent PR? The feature itself looks like standalone although this PR uses this.

Reply:
Let me do that. I included it initially here to show how it is being used. Will create a separate PR in a bit.